lib/synapse/service_watcher/zookeeper/zookeeper.rb

```diff
@@ -154,11 +171,19 @@ def watch
           @watcher = @zk.register(@discovery['path'], &watcher_callback)
         end
 
-        # Verify that we actually set up the watcher.
-        existed = with_retry(@retry_policy.merge({'retriable_errors' => ZK_RETRIABLE_ERRORS})) do |attempts|
-          log.info "synapse: zk exists at #{@discovery['path']} for #{attempts} times"
-          @zk.exists?(@discovery['path'], :watch => true)
+        # There are two cases when new watch needs to be created in zk, otherwise watch already exists.
+        # 1. Create watch in zk when zk event is nil
+        # 2. Create watch again when current event is triggered by existing watch. 
+        existed = true
+        need_exists_watch = zk_event.nil? || zk_event.node_changed? || zk_event.node_created? || zk_event.node_deleted?
```

Comment:
why does `zk_event.node_changed?` also need to set another watch?

Reply:
ZK has two types of watches, one is data watch on znode , the other is child watch on znode children. The watch has nothing to do with type of calls is made, exists() or getData(), that means this watch will be triggered when the znode is changed not deleted, so it needs to set watch again.